### PR TITLE
Improve handling of custom API Gateway options

### LINKED ISSUE #6100: Missing required key 'restApiId' in params in 1.42.0
# This is a Bug Report

## Description

* What went wrong?

After upgrading to 1.42.0, I can no longer deploy. I get:

```
  Serverless Error ---------------------------------------
 
  Missing required key 'restApiId' in params
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     OS:                     linux
     Node Version:           8.11.4
     Serverless Version:     1.42.0
```

* What did you expect should have happened?

It should have worked like the previous version did.

* What was the config you used?

<details>
<summary>`serverless.yml`</summary>

```
service: my-app-consent

cors: &cors
  origins:
    - '${file(../../packages/dev/config/env.js):env.POWER_UP_URL}'
  headers:
    - Content-Type
    - X-Amz-Date
    - Authorization
    - X-Api-Key
    - X-Amz-Security-Token
    - X-Amz-User-Agent
    - SPOOF
  allowCredentials: false

provider:
  name: aws
  runtime: nodejs8.10
  memorySize: 256
  logRetentionInDays: 14
  apiGateway:
    restApiId: '${file(../../packages/dev/config/env.js):env.REST_API_ID}'
    restApiRootResourceId: '${file(../../packages/dev/config/env.js):env.REST_API_ROOT_RESOURCE_ID}'
  deploymentBucket:
    serverSideEncryption: AES256
  environment:
    AWS_ACCOUNT_ID: '${file(../../packages/dev/config/env.js):env.AWS_ACCOUNT_ID}'
    ENV_NAMESPACE: ${self:custom.namespace}
  iamRoleStatements:
    - Effect: 'Allow'
      Resource: 'arn:aws:dynamodb:${opt:region, self:provider.region}:*:table/${self:custom.namespace}-userConsents'
      Action:
        - dynamodb:PutItem

plugins:
  - serverless-sentry
  - serverless-webpack
  - serverless-offline

package:
  individually: true

custom:
  stage: ${opt:stage, self:provider.stage}
  namespace: my-app-${self:custom.stage}
  webpack:
    webpackConfig: ./webpack.config.js
    includeModules: false
    packager: 'yarn'
  serverless-offline:
    httpsProtocol: '../../config/localhost'
    port: 8003
  sentry:
    dsn: ${file(../../packages/dev/config/env.js):env.SENTRY_FUNCTIONS_DSN}
    captureTimeoutWarnings: false

functions:
  record-consent:
    handler: src/handlers/recordConsent.lambda
    events:
      - http:
          path: /consent
          method: post
          cors: *cors
```

</details>

* What stacktrace or error message from your provider did you see?

See above

Similar or dependent issues:

None

## Additional Data

* ***Serverless Framework Version you're using***: 1.42.0
* ***Operating System***: any
* ***Stack Trace***: none
* ***Provider Error messages***: see above

### PR DESCRIPTION
**Pull request by medikoo**
## What did you implement:

It appears there are cases where we cannot reliably resolve REST API id (as it can be set directly in CloudFormation, and even resolved via CF functions, for which there are no means to resolve reliably locally).

In light of that, hack that applies custom APIGateway settings post CF deployment was updated to 
not apply any settings, if no customizations are found in `serverless.yml` and we cannot resolve REST API id.

Additionally added support for custom `apiName` (before only default resolution was assumed)

Closes #6100 

## How did you implement it:

- Added early exit if there are no APIGateway customizations and REST API id cannot be resolved.
- Added fallback for `provider.apiName` if provided

## How can we verify it:

Deploy a service that resembles documented case, and test whether it deploys as expected.

## Todos:

- [x] Write tests

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js [hunk 1]
```diff
@@ -5,6 +5,8 @@
 const _ = require('lodash');
 const BbPromise = require('bluebird');
 
+const isRestApiId = RegExp.prototype.test.bind(/^[a-z0-9]{10}$/);
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js]: Is this regex taken from AWS docs? Are we sure that the length should be 10?

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js [hunk 2]
```diff
@@ -244,4 +269,17 @@ describe('#updateStage()', () => {
       expect(context.apiGatewayRestApiId).to.equal('someRestApiId');
     });
   });
+
+  it(
+    'should not apply hack when restAPIId could not be resolved and ' +
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js]: `restApiId`

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js [hunk 3]
```diff
@@ -244,4 +269,17 @@ describe('#updateStage()', () => {
...
+      'no custom settings are applied',
+    () => {
+      context.serverless.service.provider.apiGateway = {
+        restApiId: { 'Fn::ImportValue': 'RestApiId-${self:custom.stage}' },
+      };
+      return updateStage.call(context).then(() => {
+        expect(providerRequestStub.callCount).to.equal(0);
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js]: Should we also spy on the log and see if smth. is logged to ensure that the user knows what's goind on?

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js (hunk 1 above)
Comment (medikoo) [reply lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js]: Good question. I was searching for some reference but was unable to find any.

It's just based on assumption (all ids I seen match that). Anyway I relaxed that

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js (hunk 2 above)
Comment (medikoo) [reply lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js]: Fixed

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js (hunk 3 above)
Comment (medikoo) [reply lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js]: In this case intention was to not release any warning, just succeed.

We see there's no customization to those setting sintended, so it seems safe to just exit early without a word.

When user will attempt to customize it'll receive an informative error.

Or do you feel, we should go a bit differently?

Comment (pmuens) [reply lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js]: Makes sense. Let's keep it simple :+1:

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js [hunk 4]
```diff
@@ -21,14 +23,42 @@ module.exports = {
...
+          const hasTags = Boolean(provider.tags);
+
+          if (!isTracingEnabled && !areLogsEnabled && !hasTags) {
+            // Do crash if there are no API Gateway customizations to apply
+            return null;
+          }
+
+          const errorMessage = [
+            'Rest API could not be resolved. ',
+            'This mighe be casued by a custom API Gateway setup. ',
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js]: `might`

File: lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js [hunk 5]
```diff
@@ -212,17 +212,42 @@ describe('#updateStage()', () => {
...
+        restApiId: 'foobarfoo1',
         stageName: 'dev',
       })
       .resolves({
         variables: {
           old: 'tag',
         },
       });
-    context.serverless.service.provider.apiGateway = { restApiId: 'foobar' };
+    context.serverless.service.provider.apiGateway = { restApiId: 'foobarfoo1' };
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.test.js]: Tests now should rely on `context.state`

**Pull request merged**